include/scheduler.hpp

```diff
@@ -0,0 +1,125 @@
+                    return work_return_t::ERROR;
+                } else if (result == work_return_t::INSUFFICIENT_INPUT_ITEMS) {
+                    // nothing
+                } else if (result == work_return_t::DONE) {
+                    // nothing
+                } else if (result == work_return_t::OK) {
+                    something_happened = true;
+                } else if (result == work_return_t::INSUFFICIENT_OUTPUT_ITEMS) {
+                    something_happened = true;
+                }
+            }
+            run = something_happened;
+        }
+
+        return work_return_t::DONE;
+    }
+private:
+    init_proof init;
+    fair::graph::graph &graph;
+};
+
+/**
+ * Breadth first traversal scheduler which traverses the graph starting from the source nodes in a breath first fashion
+ * detecting cycles and nodes which can be reached from several source nodes.
+ */
+class breadth_first : public node<breadth_first> {
+public:
+    explicit breadth_first(fair::graph::graph &graph) : init{fair::graph::scheduler::init(graph)} {
+        // compute the adjacency list
+        std::set<fair::graph::graph::node_model *> node_reached;
+        for (auto &e : graph.get_edges()) {
+            adjacency_list[e._src_node].insert(e._dst_node);
+            source_nodes.insert(e._src_node);
+            node_reached.insert(e._dst_node);
+        }
+        for (auto &dst : node_reached) {
+            source_nodes.erase(dst);
+        }
+    }
+
+    work_return_t work() {
+        if (!init) {
+            return work_return_t::ERROR;
+        }
+        while (true) {
+            bool anything_happened = false;
+            std::queue<fair::graph::graph::node_model *> queue{};
+            std::set<fair::graph::graph::node_model *>   reached;
+            // add all source nodes to queue
+            for (fair::graph::graph::node_model *source_node : source_nodes) {
+                queue.push(source_node);
+                reached.insert(source_node);
+            }
+            // process all nodes, adding all unvisited child nodes to the queue
+            while (!queue.empty()) {
+                fair::graph::graph::node_model *node = queue.front();
+                queue.pop();
+                auto res = node->work();
+                anything_happened |= (res == work_return_t::OK || res == work_return_t::INSUFFICIENT_OUTPUT_ITEMS);
+                if (adjacency_list.contains(node)) { // node has outgoing edges
+                    for (auto &dst : adjacency_list.at(node)) {
+                        if (!reached.contains(dst)) { // detect cycles. this could be removed if we guarantee cycle free graphs earlier
+                            queue.push(dst);
+                            reached.insert(dst);
+                        }
+                    }
+                }
+            }
+            if (!anything_happened) {
+                return work_return_t::DONE;
+            }
+        }
+    }
+
+private:
```

Comment:
member fields on top, followed by public constructor ... IMO you may skip the private: declarator since fields after the class header definition are always private.